psdisk/psdisk.c

```diff
@@ -117,199 +116,220 @@ static void psdisk_showPartsTable(void)
+{
+	uint32_t size;
+
+	if (fseek(psdisk_common.file, 0, SEEK_SET) != 0) {
+		return -1;
 	}
+	size = ptable_size(ptable->count);
 
-	memcpy(wBuff, buff, bufflen);
-	res = fwrite(wBuff, sizeof(uint8_t), psdisk_common.mem.sectorSize, psdisk_common.file);
+	if (ptable_serialize(ptable, psdisk_common.memsz, psdisk_common.blksz) < 0) {
+		return -1;
+	}
 
-	free(wBuff);
+	if (fwrite(ptable, 1, size, psdisk_common.file) != size) {
+		return -1;
+	}
 
-	return res;
+	return 0;
 }
 
 
 /* Partition table modification functions */
 
 static int psdisk_readPartsTable(void)
 {
-	int i;
-	uint32_t pCnt;
-	struct plist_node_t *pNode;
-	ptable_partition_t *pHeaders;
+	struct plist_node_t *node;
+	ptable_t *ptable;
+	uint32_t i;
+	int ret = 0;
 
-	if ((pHeaders = ptable_readPartitions(&pCnt, &psdisk_common.mem)) == NULL) {
-		fprintf(stderr,"The file contains incorrect partition table.\n");
+	ptable = psdisk_readImg();
+	if (ptable == NULL) {
+		fprintf(stderr, "The file contains incorrect partition table.\n");
 		return -1;
 	}
 
-	for (i = 0; i < pCnt; ++i) {
-		if ((pNode = (struct plist_node_t *)calloc(1, sizeof(struct plist_node_t))) == NULL) {
-			fprintf(stderr,"Cannot allocate memory, err: %s.\n", strerror(errno));
-			free(pHeaders);
-			return -1;
+	for (i = 0; i < ptable->count; i++) {
+		node = calloc(1, sizeof(struct plist_node_t));
+		if (node == NULL) {
+			fprintf(stderr, "Cannot allocate memory, err: %s.\n", strerror(errno));
+			ret = -1;
+			break;
 		}
-		pNode->pHeader = pHeaders[i];
-		LIST_INSERT_HEAD(&psdisk_common.list, pNode, ptrs);
-		++psdisk_common.pCnt;
+		node->part = ptable->parts[i];
+		LIST_INSERT_HEAD(&psdisk_common.list, node, ptrs);
+		psdisk_common.count++;
 	}
+	free(ptable);
 
-	free(pHeaders);
-
-	return 0;
+	return ret;
 }
 
 
 static int psdisk_verifyPartsTable(void)
 {
-	int i = 0;
-	uint32_t pCnt;
-	struct plist_node_t *pNode;
-	ptable_partition_t *pHeaders;
-
-	/* Verify whether data was saved correctly */
-	if ((pHeaders = ptable_readPartitions(&pCnt, &psdisk_common.mem)) == NULL) {
-		fprintf(stderr,"The file contains incorrect partition table.\n");
+	struct plist_node_t *node;
+	ptable_t *ptable;
+	int ret = 0, i = 0;
+
+	ptable = psdisk_readImg();
+	if (ptable == NULL) {
+		fprintf(stderr, "The file contains incorrect partition table.\n");
 		return -1;
 	}
 
 	/* Verify values from file after writing */
-	LIST_FOREACH(pNode, &psdisk_common.list, ptrs) {
-		if (pNode->status == part_save) {
-			if (i >= psdisk_common.pCnt || memcmp(&pNode->pHeader, &pHeaders[i], sizeof(ptable_partition_t)) != 0) {
-				free(pHeaders);
-				return -1;
+	LIST_FOREACH(node, &psdisk_common.list, ptrs) {
+		if (node->status == part_save) {
+			/* Don't check partition reserved bytes (they might get changed while saving partition table, e.g. by adding CRC) */
+			if ((i >= psdisk_common.count) || (memcmp(&node->part, &ptable->parts[i], sizeof(ptable_part_t) - sizeof(((ptable_part_t *)0)->reserved)) != 0)) {
```

Comment:
See my comment https://github.com/phoenix-rtos/phoenix-rtos-corelibs/pull/32#discussion_r1101124373 - if you change nameless one to named one, you can get header size by
`sizeof(ptable_part_t) - sizeof(((ptable_part_t *)0)->u)` (name u for an union as an example, you call call it some other name)

Most prudent way to compare this header is to compare each element individually, though, and it's imho the way to go in most cases, so your intend is obvious to the reader and code is resistant to struct layout changes

Reply:
Please note that `offsetof` is the part of C89 `stddef.h` (which is also part of the *freestanding* c environment, so it can be used literally anywhere) - so maybe we could use it here.

Reply:
`offsetof` usage seems good to me. Added in the next force-push.